controllers/scaler_controller.go

```diff
@@ -0,0 +1,309 @@
+						csvObj.Spec.InstallStrategy.StrategySpec.DeploymentSpecs[i].Spec.Replicas = &replicas
+						isUpdateRequired = true
+					}
+				}
+
+				if isUpdateRequired {
+					err = r.Client.Update(r.ctx, &csvObj)
+					if err != nil {
+						r.logger.Error(err, "failed to update csv replica", "replicas", replicas)
+						return err
+					}
+					r.logger.Info("csv updated successfully", "csvName", csvObj.Name, "replicas", replicas)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *ScalerReconciler) addWatches(req ctrl.Request) error {
+
+	if !slices.Contains(crdList, req.Name) {
+		return nil
+	}
+
+	r.logger.Info("adding dynamic watch", "object", req.Name)
+
+	kind := crdToKindMapping[req.Name]
+	err := r.addDynamicWatch(&kind)
+	if err != nil {
+		r.logger.Error(err, "failed to add dynamic watch", "object", req.Name)
+		return err
+	}
+
+	return nil
+}
+
+func (r *ScalerReconciler) addDynamicWatch(kind *metav1.PartialObjectMetadata) error {
+	return r.controller.Watch(
+		source.Kind(
+			r.mgr.GetCache(),
+			client.Object(kind),
+			&handler.EnqueueRequestForObject{},
+			predicate.Funcs{
+				// Trigger the reconcile for both creation and deletion events of the object.
+				// This ensures the replicas in the CSV are updated based on the presence or absence of the Custom Resource (CR).
+				CreateFunc: func(e event.CreateEvent) bool {
+					return true
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					return true
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					return false
+				},
+				GenericFunc: func(e event.GenericEvent) bool {
+					return false
+				},
+			},
+			predicate.GenerationChangedPredicate{},
+		),
+	)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ScalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	controller, err := ctrl.NewControllerManagedBy(mgr).
+		Named("scaler").
+		WatchesMetadata(
+			&extv1.CustomResourceDefinition{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(
+				predicate.NewPredicateFuncs(func(obj client.Object) bool {
+					return slices.Contains(crdList, obj.GetName())
+				}),
+				predicate.Funcs{
+					// Trigger a reconcile only during the creation of a specific CRD to ensure it runs exactly once for that CRD.
+					// This is required to dynamically add a watch for the corresponding Custom Resource (CR) based on the CRD name.
+					// The reconcile will be triggered with the CRD name as `req.Name`, and the reconciler will set up a watch for the CR using the CRD name.
+					CreateFunc: func(e event.CreateEvent) bool {
+						return true
+					},
+					DeleteFunc: func(e event.DeleteEvent) bool {
+						return false
+					},
+					UpdateFunc: func(e event.UpdateEvent) bool {
+						return false
+					},
+					GenericFunc: func(e event.GenericEvent) bool {
+						return false
+					},
+				},
+				predicate.GenerationChangedPredicate{},
+			),
+		).
```

Comment:
i'm just thinking, do we want to add a watch for CSV as we are effectively managing it's replicas?

Reply:
currently i dont want to because if we do that we don't leave any room for user to make the change to the replicas of the CSV also they wont be able to scale down the odf-operator because of the webhook.